packages/lib-classifier/src/components/Classifier/components/SubjectViewer/helpers/getViewerComponent/getViewerComponent.js

```diff
@@ -0,0 +1,33 @@
+import React from 'react'
+import RasterViewer from '../../components/RasterViewer'
+import convertLocationToObject from './convertLocationToObject'
+
+const viewers = {
+  image: {
+    jpeg: RasterViewer,
+    jpg: RasterViewer,
+    png: RasterViewer
+  }
+}
+
+function getViewerComponent (location) {
+  const { mimeType, mimeSubType, url } = convertLocationToObject(location)
+  let Viewer
+
+  try {
+    Viewer = viewers[mimeType][mimeSubType]
```

Comment:
Getting viewers by mime type is one of the bits of the existing classifier that was a mistake in retrospect. It's led to some fairly horrible bits of code when a subject has mixed file types in its locations array eg. audio subjects that have an mp3 file and a jpeg or png image.

Again, I'm not sure if this is a design pattern that we should be repeating in the new classifier.

Reply:
Viewers by location extension then? I can imagine that not working if there's an extension missing (if subjects are being self-hosted, for example)

Reply:
The way I imagined it working is:

- Get the viewers for each location
- Create a specific function that identifies a classifier layout by the number and type of each viewer
- Plug that layout into the UI

Reply:
Sorry, location extension won't work either. I think the whole idea of having viewers at the individual subject location level is causing us major headaches at the moment, mostly for the audio projects but possibly for text and video too. Basically, people want a UI where you can interact with more than one file, of different types, at the same time for a given subject.

Reply:
On second thoughts, the approach here works fine when you want to classify a single image, or a slideshow of images. The solution might be to have this as a basic subject viewer, good for most projects, but then have more sophisticated subject viewer components for audio and video projects.

So, somewhere at a higher level, you decide which component the current workflow is using and load that in. For example, the project owner could set the type of subject viewer component as a config option when they build the workflow.

That probably gives us more flexibility for subjects where the data is loaded as `application/csv` or `application/json` and you can interact with the viewer to control how it's rendered (eg. zooming in on a light curve for Planet Hunters.)

Reply:
But that still correlates to having a component that renders a thing and possibly provides a means of interaction with it, for each location, no? And if not, then since these viewers don't necessarily have to be rendered, we can still have another piece of logic that composes them into the desired classification interface based on the project / workflow settings.

If not, how do you imagine this working?

Reply:
Sorry, internet connection is flaking out atm - I posted the above before I saw your second comment. Which I think is advocating for a similar thing, save that I thought it could be done logically based on the location types available. I'm not so sure now...